src/sctools/fastq.py

```diff
@@ -0,0 +1,197 @@
+from collections import namedtuple
+from . import reader
+
+
+class Record:
+    """Abstract Fastq record object
+
+    Defines properties for accessing fastq record information:
+    :property name: name field
+    :property sequence: sequence field
+    :property name2: second name field
+    :property quality: quality field
+
+    Utilities for parsing read attributes:
+    :property average_quality: return the mean quality of FastqRecord
+    """
+    __slots__ = ['_data']
+
+    def __init__(self, record):
+        """
+        :param [str|bytes] record: list of four strings or bytes objects defining a
+          single fastq record
+        """
+        self._data = list(record)
+
+    @property
+    def name(self):
+        return self._data[0]
+
+    @name.setter
+    def name(self, value):
+        if not isinstance(value, (bytes, str)):
```

Comment:
Currently you can create a record with an invalid name type but setting name to an invalid type after the fact throws an error. Would it be worth having `__init__` call the setters so that this error is also thrown at record creation time? Or is there a reason why you would want to allow invalid records to be constructed?

Reply:
I can't think of a value for invalid records. I don't think calling the setters at `__init__` time will introduce a significant overhead, so I'll switch to this. Thanks for the suggestion. 👍